plugins/rts54/fu-rts54-device.c

```diff
@@ -0,0 +1,382 @@
+		.ext = FU_RTS54_EXT_WRITEFLASH,
+		.dwregaddr = GUINT32_TO_LE (addr),
+		.bufferlen = GUINT16_TO_LE (data_sz),
+		.parameters = 0,
+	};
+	guint8 buf[FU_RTS54_REPORT_LENGTH] = { 0 };
+
+	g_return_val_if_fail (data_sz <= 128, FALSE);
+	g_return_val_if_fail (data != NULL, FALSE);
+	g_return_val_if_fail (data_sz != 0, FALSE);
+
+	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
+	memcpy (buf + FU_RTS54_CMD_BUFFER_OFFSET_DATA, data, data_sz);
+	if (!fu_rts54_device_set_report (self, buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to write flash @%08x: ", (guint) addr);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54_device_verify_update_fw (FuRts54Device *self, GError **error)
+{
+	const FuRts54CmdBuffer cmd_buffer = {
+		.cmd = FU_RTS54_CMD_WRITE_DATA,
+		.ext = FU_RTS54_EXT_VERIFYUPDATE,
+		.cmd_data0 = 1,
+		.cmd_data1 = 0,
+		.cmd_data2 = 0,
+		.cmd_data3 = 0,
+		.bufferlen = GUINT16_TO_LE (1),
+		.parameters = 0,
+	};
+	guint8 buf[FU_RTS54_REPORT_LENGTH] = { 0 };
+
+	/* set then get */
+	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
+	if (!fu_rts54_device_set_report (self, buf, sizeof(buf), error))
+		return FALSE;
+	g_usleep (4 * G_USEC_PER_SEC);
+	if (!fu_rts54_device_get_report (self, buf, sizeof(buf), error))
+		return FALSE;
+
+	/* check device status */
+	if (buf[0x40] != 0x01) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_WRITE,
+				     "firmware flash failed");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_rts54_device_erase_spare_bank (FuRts54Device *self, GError **error)
+{
+	FuRts54CmdBuffer cmd_buffer = {
+		.cmd = FU_RTS54_CMD_WRITE_DATA,
+		.ext = FU_RTS54_EXT_ERASEBANK,
+		.cmd_data0 = 0,
+		.cmd_data1 = 1,
+		.cmd_data2 = 0,
+		.cmd_data3 = 0,
+		.bufferlen = 0,
+		.parameters = 0,
+	};
+	guint8 buf[FU_RTS54_REPORT_LENGTH] = { 0 };
+	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
+	if (!fu_rts54_device_set_report (self, buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to erase spare bank: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54_device_ensure_status (FuRts54Device *self, GError **error)
+{
+	const FuRts54CmdBuffer cmd_buffer = {
+		.cmd = FU_RTS54_CMD_READ_DATA,
+		.ext = FU_RTS54_EXT_READ_STATUS,
+		.cmd_data0 = 0,
+		.cmd_data1 = 0,
+		.cmd_data2 = 0,
+		.cmd_data3 = 0,
+		.bufferlen = GUINT16_TO_LE (32),
+		.parameters = 0,
+	};
+	guint8 buf[FU_RTS54_REPORT_LENGTH] = { 0 };
+	g_autofree gchar *version = NULL;
+
+	/* set then get */
+	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
+	if (!fu_rts54_device_set_report (self, buf, sizeof(buf), error))
+		return FALSE;
+	if (!fu_rts54_device_get_report (self, buf, sizeof(buf), error))
+		return FALSE;
+
+	/* check the hardware capabilities */
+	self->dual_bank = (buf[0x40 + 7] & 0xf0) == 0x80;
+	self->fw_auth = (buf[0x40 + 13] & 0x02) > 0;
+	self->status1 = buf[0x40 + 25];
+	self->status2 = buf[0x40 + 26];
```

Comment:
@AnyProblem do any of these bits indicate the ability to communicate to I2C bridge?  So that you could differentiate between a hub with or without I2C bridge function?